File: java/word-count/src/main/java/WordCount.java
import java.util.HashMap;
import java.util.Map;

public class WordCount {
	
	private static final String NOT_ALPHA_NUMERIC = "[^a-zA-Z0-9']+";
	private static final String WHITESPACE = " ";

	public Map<String, Integer> Phrase(String phrase) {

		Map<String, Integer> wordCount = new HashMap<>();
		
		String[] words = phrase.replaceAll(NOT_ALPHA_NUMERIC, WHITESPACE).toLowerCase().split(WHITESPACE);
		
		for (String word : words) {
			
			if (wordCount.containsKey(word)) {
				int newCount = wordCount.get(word) + 1;
				wordCount.put(word, newCount);
			
			} else {
				wordCount.put(word, 1);
			}
		}
		return wordCount;
	}
}

File: ecmascript/word-count/word-count.js
export default class Words {
  count(phrase) {
    const words = phrase.trim().toLowerCase().split(/\s+/);
    return words.reduce((wordCount, word) => containsKey(wordCount, word)
      ? putKeyValue(wordCount, word, wordCount[word] + 1)
      : putKeyValue(wordCount, word, 1),
      {}
    );
  }
}

const containsKey = (obj, key) => Object.keys(obj).includes(key);
const putKeyValue = (obj, key, value) => ({ ...obj, [key]: value });
